core/ledger/kvledger/txmgmt/statedb/statecouchdb/statecouchdb_test.go

```diff
@@ -662,7 +662,7 @@ func TestHandleChaincodeDeployErroneousIndexFile(t *testing.T) {
 		}
 		return true
 	}
-	assert.Eventually(t, queryUsingIndex, 2*time.Second, 100*time.Millisecond, "error executing query with sort")
```

Comment:
This is the assertion failure that I'm seeing in CI. I see that there was another change for `TestHandleChaincodeDeploy` but I don't know if that' was due to an actual failure.

```
2020-04-17T14:17:09.8628824Z --- FAIL: TestHandleChaincodeDeployErroneousIndexFile (2.30s)
2020-04-17T14:17:09.8629124Z     statecouchdb_test.go:39: Initializing TestVDBEnv
2020-04-17T14:17:09.8629382Z     statecouchdb_test.go:741: 
2020-04-17T14:17:09.8629650Z         	Error Trace:	statecouchdb_test.go:741
2020-04-17T14:17:09.8629922Z         	Error:      	Condition never satisfied
2020-04-17T14:17:09.8630206Z         	Test:       	TestHandleChaincodeDeployErroneousIndexFile
2020-04-17T14:17:09.8630518Z         	Messages:   	error executing query with sort
2020-04-17T14:17:09.8630796Z     statecouchdb_test.go:86: Cleaningup TestVDBEnv
```

Focusing on this test, and to be fair, I'm not really sure what the intent is...

The name implies we want to see what happens when a bad index file is used, asserts there's an error creating the indexes, and then there's a query that actually depends on the creation of the index from `indexSizeSortName.json`.

The problem is that `ProcessIndexesForChaincodeDeploy` is iterating over the map of `indexData` and returning on the first error. As map iteration is non-deterministic, there's a 50/50 shot this the first index is the bad one. When that happens, the second index isn't created and the assertion in eventually will never succeed.

So, to me, the fix would be to make this deterministic. One way is to process the "good" index first, assert no errors, and then do the bad.

Reply:
My local change, for example, successfully runs 100 iterations of the test:

```
✓ [11:30:35] [~/workspace/fabric/src/github.com/hyperledger/fabric]
[sykesm@trapped] (master *)$ go test -run TestHandleChaincodeDeployErroneousIndexFile -failfast -count 100 ./core/ledger/kvledger/txmgmt/statedb/statecouchdb/
ok  	github.com/hyperledger/fabric/core/ledger/kvledger/txmgmt/statedb/statecouchdb	33.235s
```

```diff
diff --git a/core/ledger/kvledger/txmgmt/statedb/statecouchdb/statecouchdb_test.go b/core/ledger/kvledger/txmgmt/statedb/statecouchdb/statecouchdb_test.go
index 58530ca19..435091b00 100644
--- a/core/ledger/kvledger/txmgmt/statedb/statecouchdb/statecouchdb_test.go
+++ b/core/ledger/kvledger/txmgmt/statedb/statecouchdb/statecouchdb_test.go
@@ -718,18 +718,21 @@ func TestHandleChaincodeDeployErroneousIndexFile(t *testing.T) {
        batch.Put("ns1", "key1", []byte(`{"asset_name": "marble1","color": "blue","size": 1,"owner": "tom"}`), version.NewHeight(1, 1))
        batch.Put("ns1", "key2", []byte(`{"asset_name": "marble2","color": "blue","size": 2,"owner": "jerry"}`), version.NewHeight(1, 2))
 
-       badSyntaxFileContent := `{"index":{"fields": This is a bad json}`
        indexData := map[string][]byte{
                "META-INF/statedb/couchdb/indexes/indexSizeSortName.json": []byte(`{"index":{"fields":[{"size":"desc"}]},"ddoc":"indexSizeSortName","name":"indexSizeSortName","type":"json"}`),
-               "META-INF/statedb/couchdb/indexes/badSyntax.json":         []byte(badSyntaxFileContent),
        }
 
        indexCapable, ok := db.(statedb.IndexCapable)
        if !ok {
                t.Fatalf("Couchdb state impl is expected to implement interface `statedb.IndexCapable`")
        }
-       assert.Error(t, indexCapable.ProcessIndexesForChaincodeDeploy("ns1", indexData))
+       assert.NoError(t, indexCapable.ProcessIndexesForChaincodeDeploy("ns1", indexData))
 
+       badSyntaxFileContent := `{"index":{"fields": This is a bad json}`
+       indexData = map[string][]byte{
+               "META-INF/statedb/couchdb/indexes/badSyntax.json": []byte(badSyntaxFileContent),
+       }
+       assert.Error(t, indexCapable.ProcessIndexesForChaincodeDeploy("ns1", indexData))
        queryString := `{"selector":{"owner":"fred"}, "sort": [{"size": "desc"}]}`
        queryUsingIndex := func() bool {
                _, err = db.ExecuteQuery("ns1", queryString)
```

Reply:
Thanks, @sykesm for teaching me how to do it correctly. We have to be careful with the maps. This non-determinism was introduced by #966 As you have the fix, can you please revert the changes I made and push yours?

@btl5037 Sorry for the trouble. I had a wrong hypothesis on the resource usage.

Reply:
@sykesm This could create non-determinism even in the production setup. Some peers would have the index and some wouldn't. Hence, let me discuss this issue in ledger scrum today and push a fix by tonight.